src/backend/distributed/metadata/node_metadata.c

```diff
@@ -388,13 +391,13 @@ SetUpDistributedTableDependencies(WorkerNode *newWorkerNode)
 	{
 		EnsureNoModificationsHaveBeenDone();
 
-		if (ShouldPropagate())
+		if (ShouldPropagate() && !NodeIsCoordinator(newWorkerNode))
```

Comment:
Just as an opinion, I think that here checking NodeIsCoordinator seems a bit weird to me. We don't want to propagate the dependencies for a node that already has the dependencies. Currently this can only be the coordinator because we enable running a command that would trigger this only on the coordinator. However in the future, we might enable MX nodes as well to be able to run such commands, and in that case we will have the same problem. So to me the better check is that we check if the newWorkerNode already had the dependencies e.g it was already in the cluster.

Reply:
There are a number of things that we can not do on MX nodes, and altering metadata is one of them. If we want to add node, master_add_node() can only be called from the coordinator for now, and as you pointed out it can be changed in the future.

Currently if you run master_add_node with an existing hostname and port that has a record in pg_dist_node, we early exit and just return the id of the pg_dist_node record. I can not think of a scenario where it can be beneficial to add an MX node multiple times, and not early exit.